File: IOS/ALAFIA/Models/UserRoles.swift
import Foundation

// MARK: - Role Catalog

struct RoleCatalogEntry: Codable, Identifiable {
    let id: String
    let name: String
    let category: String
    let description: String?
}

struct RoleCategoryInfo: Codable, Identifiable {
    let id: String
    let name: String
    let roles: [RoleCatalogEntry]
}

struct RoleCategorySummary: Codable, Identifiable {
    let id: String
    let name: String
    let count: Int
}

// MARK: - Professional Profile

struct ProfessionalProfile: Codable, Identifiable {
    let id: Int
    let roleAssignmentId: Int
    // Credentials
    let licenseNumber: String?
    let licenseState: String?
    let licenseCountry: String?
    let licenseExpiry: String?
    let npiNumber: String?
    let deaNumber: String?
    let boardCertifications: [String]?
    // Education
    let medicalSchool: String?
    let degree: String?
    let graduationYear: Int?
    let residencyProgram: String?
    let fellowshipProgram: String?
    // Practice
    let specialty: String?
    let subSpecialty: String?
    let yearsOfExperience: Int?
    let practiceName: String?
    let practiceType: String?
    let practiceAddress: String?
    let practicePhone: String?
    let practiceEmail: String?
    let practiceWebsite: String?
    let acceptingPatients: Bool?
    // Hospital
    let hospitalAffiliations: [String]?
    let department: String?
    let title: String?
    // Languages
    let clinicalLanguages: [String]?
    // Telemedicine
    let telemedicineAvailable: Bool?
    let telemedicinePlatforms: [String]?
    // Verification
    let verificationStatus: String?
    let verificationDate: String?
    // Bio
    let professionalBio: String?
    let publications: [String]?
    let researchInterests: [String]?
    
    let createdAt: String?
    let updatedAt: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case roleAssignmentId = "role_assignment_id"
        case licenseNumber = "license_number"
        case licenseState = "license_state"
        case licenseCountry = "license_country"
        case licenseExpiry = "license_expiry"
        case npiNumber = "npi_number"
        case deaNumber = "dea_number"
        case boardCertifications = "board_certifications"
        case medicalSchool = "medical_school"
        case degree
        case graduationYear = "graduation_year"
        case residencyProgram = "residency_program"
        case fellowshipProgram = "fellowship_program"
        case specialty
        case subSpecialty = "sub_specialty"
        case yearsOfExperience = "years_of_experience"
        case practiceName = "practice_name"
        case practiceType = "practice_type"
        case practiceAddress = "practice_address"
        case practicePhone = "practice_phone"
        case practiceEmail = "practice_email"
        case practiceWebsite = "practice_website"
        case acceptingPatients = "accepting_patients"
        case hospitalAffiliations = "hospital_affiliations"
        case department, title
        case clinicalLanguages = "clinical_languages"
        case telemedicineAvailable = "telemedicine_available"
        case telemedicinePlatforms = "telemedicine_platforms"
        case verificationStatus = "verification_status"
        case verificationDate = "verification_date"
        case professionalBio = "professional_bio"
        case publications
        case researchInterests = "research_interests"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

// MARK: - Role Assignment

struct RoleAssignment: Codable, Identifiable {
    let id: Int
    let userId: Int
    let role: String
    let isPrimary: Bool
    let isActive: Bool
    let grantedAt: String?
    let professionalProfile: ProfessionalProfile?
    
    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case role
        case isPrimary = "is_primary"
        case isActive = "is_active"
        case grantedAt = "granted_at"
        case professionalProfile = "professional_profile"
    }
    
    var displayName: String {
        role.replacingOccurrences(of: "_", with: " ").capitalized
    }
}

// MARK: - User Persona Summary

struct UserPersonaSummary: Codable {
    let userId: Int
    let fullName: String
    let email: String
    let primaryRole: String
    let activeRoles: [String]
    let roleDetails: [RoleAssignment]
    let isHealthcareProfessional: Bool
    let roleCategories: [String]
    let permissions: [String]
    
    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case fullName = "full_name"
        case email
        case primaryRole = "primary_role"
        case activeRoles = "active_roles"
        case roleDetails = "role_details"
        case isHealthcareProfessional = "is_healthcare_professional"
        case roleCategories = "role_categories"
        case permissions
    }
}

// MARK: - Request Models

struct RoleAssignmentCreate: Encodable {
    let role: String
    let isPrimary: Bool
    
    enum CodingKeys: String, CodingKey {
        case role
        case isPrimary = "is_primary"
    }
}

struct ProfessionalProfileCreate: Encodable {
    var licenseNumber: String?
    var licenseState: String?
    var licenseCountry: String?
    var licenseExpiry: String?
    var npiNumber: String?
    var deaNumber: String?
    var boardCertifications: [String]?
    var medicalSchool: String?
    var degree: String?
    var graduationYear: Int?
    var residencyProgram: String?
    var fellowshipProgram: String?
    var specialty: String?
    var subSpecialty: String?
    var yearsOfExperience: Int?
    var practiceName: String?
    var practiceType: String?
    var practiceAddress: String?
    var practicePhone: String?
    var practiceEmail: String?
    var practiceWebsite: String?
    var acceptingPatients: Bool?
    var hospitalAffiliations: [String]?
    var department: String?
    var title: String?
    var clinicalLanguages: [String]?
    var telemedicineAvailable: Bool?
    var telemedicinePlatforms: [String]?
    var professionalBio: String?
    var publications: [String]?
    var researchInterests: [String]?
    
    enum CodingKeys: String, CodingKey {
        case licenseNumber = "license_number"
        case licenseState = "license_state"
        case licenseCountry = "license_country"
        case licenseExpiry = "license_expiry"
        case npiNumber = "npi_number"
        case deaNumber = "dea_number"
        case boardCertifications = "board_certifications"
        case medicalSchool = "medical_school"
        case degree
        case graduationYear = "graduation_year"
        case residencyProgram = "residency_program"
        case fellowshipProgram = "fellowship_program"
        case specialty
        case subSpecialty = "sub_specialty"
        case yearsOfExperience = "years_of_experience"
        case practiceName = "practice_name"
        case practiceType = "practice_type"
        case practiceAddress = "practice_address"
        case practicePhone = "practice_phone"
        case practiceEmail = "practice_email"
        case practiceWebsite = "practice_website"
        case acceptingPatients = "accepting_patients"
        case hospitalAffiliations = "hospital_affiliations"
        case department, title
        case clinicalLanguages = "clinical_languages"
        case telemedicineAvailable = "telemedicine_available"
        case telemedicinePlatforms = "telemedicine_platforms"
        case professionalBio = "professional_bio"
        case publications
        case researchInterests = "research_interests"
    }
}

// MARK: - Category Icons

let roleCategoryIcons: [String: String] = [
    "physician": "stethoscope",
    "surgeon": "scissors",
    "nursing": "cross.case.fill",
    "therapy": "figure.mind.and.body",
    "mental_health": "brain.head.profile",
    "social_work": "person.2.fill",
    "dental": "mouth.fill",
    "pharmacy": "pills.fill",
    "diagnostics": "microscope",
    "emergency": "cross.circle.fill",
    "allied_health": "heart.text.clipboard",
    "public_health": "globe.americas.fill",
    "administration": "building.2.fill",
    "research": "magnifyingglass",
    "traditional_complementary": "leaf.fill",
    "midwifery": "figure.and.child.holdinghands",
    "other": "person.fill.questionmark",
]

let verificationStatusColors: [String: String] = [
    "unverified": "gray",
    "pending": "orange",
    "verified": "green",
    "rejected": "red",
    "expired": "secondary",
]

File: WEB/backend/app/schemas/user_roles.py
"""Schemas for user roles and professional profiles."""

from pydantic import BaseModel
from typing import Optional
from datetime import datetime


# ── Role Assignment ───────────────────────────────────────────────────

class RoleAssignmentCreate(BaseModel):
    role: str
    is_primary: bool = False


class RoleAssignmentResponse(BaseModel):
    id: int
    user_id: int
    role: str
    is_primary: bool
    is_active: bool
    granted_at: datetime | None = None
    professional_profile: Optional["ProfessionalProfileResponse"] = None

    model_config = {"from_attributes": True}


# ── Professional Profile ─────────────────────────────────────────────

class ProfessionalProfileCreate(BaseModel):
    # Credentials
    license_number: str | None = None
    license_state: str | None = None
    license_country: str | None = None
    license_expiry: str | None = None
    npi_number: str | None = None
    dea_number: str | None = None
    board_certifications: list[str] | None = None

    # Education
    medical_school: str | None = None
    degree: str | None = None
    graduation_year: int | None = None
    residency_program: str | None = None
    fellowship_program: str | None = None

    # Practice
    specialty: str | None = None
    sub_specialty: str | None = None
    years_of_experience: int | None = None
    practice_name: str | None = None
    practice_type: str | None = None
    practice_address: str | None = None
    practice_phone: str | None = None
    practice_email: str | None = None
    practice_website: str | None = None
    accepting_patients: bool | None = None

    # Hospital
    hospital_affiliations: list[str] | None = None
    department: str | None = None
    title: str | None = None

    # Languages
    clinical_languages: list[str] | None = None

    # Telemedicine
    telemedicine_available: bool = False
    telemedicine_platforms: list[str] | None = None

    # Bio
    professional_bio: str | None = None
    publications: list[str] | None = None
    research_interests: list[str] | None = None


class ProfessionalProfileUpdate(BaseModel):
    license_number: str | None = None
    license_state: str | None = None
    license_country: str | None = None
    license_expiry: str | None = None
    npi_number: str | None = None
    dea_number: str | None = None
    board_certifications: list[str] | None = None
    medical_school: str | None = None
    degree: str | None = None
    graduation_year: int | None = None
    residency_program: str | None = None
    fellowship_program: str | None = None
    specialty: str | None = None
    sub_specialty: str | None = None
    years_of_experience: int | None = None
    practice_name: str | None = None
    practice_type: str | None = None
    practice_address: str | None = None
    practice_phone: str | None = None
    practice_email: str | None = None
    practice_website: str | None = None
    accepting_patients: bool | None = None
    hospital_affiliations: list[str] | None = None
    department: str | None = None
    title: str | None = None
    clinical_languages: list[str] | None = None
    telemedicine_available: bool | None = None
    telemedicine_platforms: list[str] | None = None
    professional_bio: str | None = None
    publications: list[str] | None = None
    research_interests: list[str] | None = None


class ProfessionalProfileResponse(BaseModel):
    id: int
    role_assignment_id: int

    # Credentials
    license_number: str | None = None
    license_state: str | None = None
    license_country: str | None = None
    license_expiry: str | None = None
    npi_number: str | None = None
    dea_number: str | None = None
    board_certifications: list[str] | None = None

    # Education
    medical_school: str | None = None
    degree: str | None = None
    graduation_year: int | None = None
    residency_program: str | None = None
    fellowship_program: str | None = None

    # Practice
    specialty: str | None = None
    sub_specialty: str | None = None
    years_of_experience: int | None = None
    practice_name: str | None = None
    practice_type: str | None = None
    practice_address: str | None = None
    practice_phone: str | None = None
    practice_email: str | None = None
    practice_website: str | None = None
    accepting_patients: bool | None = None

    # Hospital
    hospital_affiliations: list[str] | None = None
    department: str | None = None
    title: str | None = None

    # Languages
    clinical_languages: list[str] | None = None

    # Telemedicine
    telemedicine_available: bool = False
    telemedicine_platforms: list[str] | None = None

    # Verification
    verification_status: str = "unverified"
    verification_date: datetime | None = None

    # Bio
    professional_bio: str | None = None
    publications: list[str] | None = None
    research_interests: list[str] | None = None

    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


# ── User Persona Summary ─────────────────────────────────────────────

class UserPersonaSummary(BaseModel):
    """Returned by /users/me — enriched with persona info."""
    user_id: int
    full_name: str
    email: str
    primary_role: str  # always at least "patient"
    active_roles: list[str]
    role_details: list[RoleAssignmentResponse]
    is_healthcare_professional: bool
    role_categories: list[str]  # e.g. ["physician", "administration"]
    permissions: list[str]  # derived feature-access list


# ── Role Catalog (reference data) ────────────────────────────────────

class RoleCatalogEntry(BaseModel):
    id: str
    name: str
    category: str
    description: str | None = None


class RoleCategoryInfo(BaseModel):
    id: str
    name: str
    roles: list[RoleCatalogEntry]
